clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -858,6 +861,55 @@ class KernelObjVisitor {
   void VisitRecord(const CXXRecordDecl *Owner, ParentTy &Parent,
                    const CXXRecordDecl *Wrapper, Handlers &... handlers);
 
+  // Base case, only calls these when filtered.
+  template <typename... FilteredHandlers, typename ParentTy>
+  void VisitUnion(FilteredHandlers &... handlers, const CXXRecordDecl *Owner,
+                  ParentTy &Parent, const CXXRecordDecl *Wrapper) {
+    (void)std::initializer_list<int>{
+        (handlers.enterUnion(Owner, Parent), 0)...};
+    VisitRecordHelper(Wrapper, Wrapper->fields(), handlers...);
+    (void)std::initializer_list<int>{
+        (handlers.leaveUnion(Owner, Parent), 0)...};
+  }
+
+  template <typename... FilteredHandlers, typename ParentTy,
+            typename CurHandler, typename... Handlers>
+  std::enable_if_t<!CurHandler::VisitUnionBody>
+  VisitUnion(FilteredHandlers &... filtered_handlers,
+             const CXXRecordDecl *Owner, ParentTy &Parent,
+             const CXXRecordDecl *Wrapper, CurHandler &cur_handler,
+             Handlers &... handlers) {
+    VisitUnion<FilteredHandlers...>(filtered_handlers..., Owner, Parent,
+                                    Wrapper, handlers...);
+  }
+
+  template <typename... FilteredHandlers, typename ParentTy,
+            typename CurHandler, typename... Handlers>
+  std::enable_if_t<CurHandler::VisitUnionBody>
+  VisitUnion(FilteredHandlers &... filtered_handlers,
+             const CXXRecordDecl *Owner, ParentTy &Parent,
+             const CXXRecordDecl *Wrapper, CurHandler &cur_handler,
+             Handlers &... handlers) {
+    VisitUnion<FilteredHandlers..., CurHandler>(
+        filtered_handlers..., cur_handler, Owner, Parent, Wrapper, handlers...);
+  }
+
+  template <typename ParentTy, typename CurHandler, typename... Handlers>
```

Comment:
For everyone else's edification: The MSVC compiler has trouble with leading-empty packs, so we need to have these two overloads that are identical to the above two, except they don't have filtered-handlers.  It just makes these the entry-point, which work exactly the same as the other two.